File: python_bot/src/account_service.py
import logging
from decimal import Decimal
from typing import List, Optional
from .models import Account, Position, Currency

logger = logging.getLogger(__name__)


class AccountService:
    """Wraps Wealthsimple account-related endpoints."""

    def __init__(self, client):
        self._client = client

    def get_accounts(self) -> List[Account]:
        data = self._client.get("/account/list")
        results = data.get("results", [])
        accounts = []
        for raw in results:
            try:
                account = Account(
                    id=raw["id"],
                    account_type=raw.get("account_type", ""),
                    buying_power=Currency(
                        amount=Decimal(str(raw.get("buying_power", {}).get("amount", 0))),
                        currency=raw.get("buying_power", {}).get("currency", "CAD"),
                    ),
                    current_balance=Currency(
                        amount=Decimal(str(raw.get("current_balance", {}).get("amount", 0))),
                        currency=raw.get("current_balance", {}).get("currency", "CAD"),
                    ),
                    net_deposits=Currency(
                        amount=Decimal(str(raw.get("net_deposits", {}).get("amount", 0))),
                        currency=raw.get("net_deposits", {}).get("currency", "CAD"),
                    ),
                    status=raw.get("status", ""),
                )
                accounts.append(account)
            except (KeyError, TypeError) as e:
                logger.warning("Failed to parse account: %s — %s", raw.get("id", "?"), e)
        logger.info("Found %d accounts", len(accounts))
        return accounts

    def get_account_by_type(self, account_type: str) -> Optional[Account]:
        accounts = self.get_accounts()
        for acc in accounts:
            if acc.account_type == account_type:
                return acc
        logger.warning("No account found with type: %s", account_type)
        return None

    def get_positions(self, account_id: Optional[str] = None) -> List[Position]:
        params = {}
        if account_id:
            params["account_id"] = account_id
        data = self._client.get("/account/positions", params=params)
        results = data.get("results", [])
        positions = []
        for raw in results:
            try:
                stock = raw.get("stock", {})
                quote = raw.get("quote", {})
                quantity = Decimal(str(raw.get("quantity", 0)))
                current_price = Decimal(str(quote.get("amount", 0)))
                market_value = quantity * current_price
                book_value = Decimal(str(raw.get("book_value", {}).get("amount", 0)))

                pos = Position(
                    security_id=raw.get("id", raw.get("security_id", "")),
                    symbol=stock.get("symbol", ""),
                    quantity=quantity,
                    market_value=market_value,
                    book_value=book_value,
                    currency=stock.get("currency", "CAD"),
                    average_price=Decimal(str(raw.get("entry_price", {}).get("amount", 0))),
                    current_price=current_price,
                    gain_loss=market_value - book_value,
                    gain_loss_pct=(
                        ((market_value - book_value) / book_value * 100)
                        if book_value > 0
                        else Decimal("0")
                    ),
                )
                positions.append(pos)
            except (KeyError, TypeError, ZeroDivisionError) as e:
                logger.warning("Failed to parse position: %s", e)
        logger.info("Found %d positions", len(positions))
        return positions

    def get_buying_power(self, account_id: str) -> Decimal:
        accounts = self.get_accounts()
        for acc in accounts:
            if acc.id == account_id:
                return acc.buying_power.amount
        return Decimal("0")

    def get_total_portfolio_value(
        self, account_id: str, positions: Optional[List[Position]] = None
    ) -> Decimal:
        cash = self.get_buying_power(account_id)
        if positions is None:
            positions = self.get_positions(account_id)
        positions_value = sum(p.market_value for p in positions)
        total = cash + positions_value
        logger.info(
            "Portfolio value: cash=$%.2f + positions=$%.2f = total=$%.2f",
            cash, positions_value, total,
        )
        return total

File: dotnet_bot/src/WealthsimpleTradingBot/Api/AccountService.cs
using System.Text.Json;
using WealthsimpleTradingBot.Models;

namespace WealthsimpleTradingBot.Api;

public interface IAccountService
{
    Task<List<Account>> GetAccountsAsync();
    Task<Account?> GetAccountByTypeAsync(string accountType);
    Task<List<Position>> GetPositionsAsync(string? accountId = null);
    Task<decimal> GetBuyingPowerAsync(string accountId);
    Task<decimal> GetTotalPortfolioValueAsync(string accountId, List<Position>? positions = null);
}

public class AccountService : IAccountService
{
    private readonly IWealthsimpleClient _client;
    private readonly ILogger<AccountService> _logger;

    public AccountService(IWealthsimpleClient client, ILogger<AccountService> logger)
    {
        _client = client;
        _logger = logger;
    }

    public async Task<List<Account>> GetAccountsAsync()
    {
        var data = await _client.GetAsync<JsonElement>("/account/list");
        var accounts = new List<Account>();

        if (data.TryGetProperty("results", out var results))
        {
            foreach (var raw in results.EnumerateArray())
            {
                try
                {
                    accounts.Add(new Account(
                        Id: raw.GetProperty("id").GetString()!,
                        AccountType: GetStringProp(raw, "account_type"),
                        BuyingPower: ParseCurrency(raw, "buying_power"),
                        CurrentBalance: ParseCurrency(raw, "current_balance"),
                        NetDeposits: ParseCurrency(raw, "net_deposits"),
                        Status: GetStringProp(raw, "status")
                    ));
                }
                catch (Exception ex)
                {
                    _logger.LogWarning("Failed to parse account: {Error}", ex.Message);
                }
            }
        }

        _logger.LogInformation("Found {Count} accounts", accounts.Count);
        return accounts;
    }

    public async Task<Account?> GetAccountByTypeAsync(string accountType)
    {
        var accounts = await GetAccountsAsync();
        var account = accounts.FirstOrDefault(a => a.AccountType == accountType);
        if (account == null)
            _logger.LogWarning("No account found with type: {Type}", accountType);
        return account;
    }

    public async Task<List<Position>> GetPositionsAsync(string? accountId = null)
    {
        var queryParams = accountId != null
            ? new Dictionary<string, string> { ["account_id"] = accountId }
            : null;

        var data = await _client.GetAsync<JsonElement>("/account/positions", queryParams);
        var positions = new List<Position>();

        if (data.TryGetProperty("results", out var results))
        {
            foreach (var raw in results.EnumerateArray())
            {
                try
                {
                    var quantity = GetDecimalProp(raw, "quantity");
                    var currentPrice = GetNestedDecimal(raw, "quote", "amount");
                    var marketValue = quantity * currentPrice;
                    var bookValue = GetNestedDecimal(raw, "book_value", "amount");

                    positions.Add(new Position(
                        SecurityId: GetStringProp(raw, "id"),
                        Symbol: GetNestedString(raw, "stock", "symbol"),
                        Quantity: quantity,
                        MarketValue: marketValue,
                        BookValue: bookValue,
                        Currency: GetNestedString(raw, "stock", "currency"),
                        AveragePrice: GetNestedDecimal(raw, "entry_price", "amount"),
                        CurrentPrice: currentPrice,
                        GainLoss: marketValue - bookValue,
                        GainLossPct: bookValue > 0 ? (marketValue - bookValue) / bookValue * 100 : 0
                    ));
                }
                catch (Exception ex)
                {
                    _logger.LogWarning("Failed to parse position: {Error}", ex.Message);
                }
            }
        }

        _logger.LogInformation("Found {Count} positions", positions.Count);
        return positions;
    }

    public async Task<decimal> GetBuyingPowerAsync(string accountId)
    {
        var accounts = await GetAccountsAsync();
        return accounts.FirstOrDefault(a => a.Id == accountId)?.BuyingPower.Amount ?? 0;
    }

    public async Task<decimal> GetTotalPortfolioValueAsync(
        string accountId, List<Position>? positions = null)
    {
        var cash = await GetBuyingPowerAsync(accountId);
        positions ??= await GetPositionsAsync(accountId);
        var positionsValue = positions.Sum(p => p.MarketValue);
        var total = cash + positionsValue;
        _logger.LogInformation(
            "Portfolio: cash={Cash:C}, positions={Positions:C}, total={Total:C}",
            cash, positionsValue, total);
        return total;
    }

    private static CurrencyAmount ParseCurrency(JsonElement parent, string propName)
    {
        if (parent.TryGetProperty(propName, out var curr))
        {
            return new CurrencyAmount(
                Amount: curr.TryGetProperty("amount", out var amt) ? amt.GetDecimal() : 0,
                Currency: curr.TryGetProperty("currency", out var c) ? c.GetString()! : "CAD"
            );
        }
        return new CurrencyAmount(0, "CAD");
    }

    private static string GetStringProp(JsonElement el, string name)
        => el.TryGetProperty(name, out var val) ? val.GetString() ?? "" : "";

    private static decimal GetDecimalProp(JsonElement el, string name)
        => el.TryGetProperty(name, out var val) ? val.GetDecimal() : 0;

    private static string GetNestedString(JsonElement el, string parent, string child)
    {
        if (el.TryGetProperty(parent, out var p) && p.TryGetProperty(child, out var c))
            return c.GetString() ?? "";
        return "";
    }

    private static decimal GetNestedDecimal(JsonElement el, string parent, string child)
    {
        if (el.TryGetProperty(parent, out var p) && p.TryGetProperty(child, out var c))
            return c.GetDecimal();
        return 0;
    }
}
